frame/transaction-storage/src/lib.rs

```diff
@@ -0,0 +1,364 @@
+	CHUNK_SIZE, INHERENT_IDENTIFIER, DEFAULT_STORAGE_PERIOD,
+};
+
+/// A type alias for the balance type from this pallet's point of view.
+type BalanceOf<T> = <<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+// Re-export pallet items so that they can be accessed from the crate namespace.
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+/// Maximum bytes that can be storead in one transaction.
+// Increasing it further also requires raising the allocator limit.
+pub const MAX_DATA_SIZE: u32 = 8 * 1024 * 1024;
+
+/// State data for a stored transaction.
+#[derive(Encode, Decode, sp_runtime::RuntimeDebug, PartialEq, Eq)]
+pub struct TransactionInfo {
+	/// Chunk trie root.
+	chunk_root: <BlakeTwo256 as Hash>::Output,
+	/// Plain hash of indexed data.
+	content_hash: <BlakeTwo256 as Hash>::Output,
+	/// Size of indexed data in bytes.
+	size: u32,
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// A dispatchable call.
+		type Call: Parameter + Dispatchable<Origin=Self::Origin> + GetDispatchInfo + From<frame_system::Call<Self>>;
+		/// The currency trait.
+		type Currency: ReservableCurrency<Self::AccountId>;
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Insufficient account balance.
+		InsufficientFunds,
+		/// Invalid configuration.
+		NotConfigured,
+		/// Renewed extrinsic is not found.
+		RenewedNotFound,
+		/// Attempting to store empty transaction
+		EmptyTransaction,
+		/// Proof was not expected in this block.
+		UnexpectedProof,
+		/// Proof failed verification.
+		InvalidProof,
+		/// Missing storage proof.
+		MissingProof,
+		/// Unable to verify proof becasue state data is missing.
+		MissingStateData,
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			let period = <StoragePeriod<T>>::get();
+			let obsolete = n.saturating_sub(period);
+			let transaction_count = if obsolete > Zero::zero() {
+				<TransactionCount<T>>::get(obsolete)
+			} else {
+				0
+			};
+			// return weight for `on_finalize` does
+			T::WeightInfo::on_finalize(transaction_count)
+		}
+
+		fn on_finalize(n: T::BlockNumber) {
+			<TransactionCount<T>>::insert(n, <Counter<T>>::get());
+			<Counter<T>>::kill();
+			// Drop obsolete roots
+			let period = <StoragePeriod<T>>::get();
+			let obsolete = n.saturating_sub(period);
+			if obsolete > Zero::zero() {
+				<TransactionRoots<T>>::remove_prefix(obsolete);
+				<TransactionCount<T>>::remove(obsolete);
```

Comment:
is there an alternative to doing this in the `on_finalize`?

I see that you do account for this properly in the `on_initialize` weight, but even still, we really want to avoid logic in `on_finalize` if we can avoid it.

Reply:
For example, if we are just cleaning things up, why not clean it up at the beginning of the block?

Reply:
> we really want to avoid logic in on_finalize if we can avoid it.

Why? If I don't clean up temporary values in `on_finalize` they'll end up being committed to the database. Something I'd like to avoid.

Reply:
Ah, I can cleanup obsolete data in `on_initialize` indeed.

I'm still curious why `on_initalize` is prefered to  `on_finalize`? Until very recently, `on_initalize` was needed to be called before executing not just blocks, but some runtime APIs. So i'd expected we'd want to put as little as possible there.

Reply:
@arkpar on initialize is preferred because it all happens at the beginning of the block before transactions are included.

This means that for things like PoV size, we will know at the beginning of a block, before we add txs, how much is used and that most probably it will fit because there is nothing else used in the space.

The runtime can handle situations where transactions are not included, but the runtime CANNOT handle not including the `on_finalize` or `on_initialize`. So while the block is empty and we can most safely do all operations, we want to stick it in the front.